### ISSUE TITLE
Allow continue compressing after abortion

### ISSUE THREAD

**Issue by LuckyTurtleDev** [shssoichiro/oxipng#549]
If the user has a slow pc or using a high compression, he might can not compress all images at once.
But after aborting it is not possible to continue at the last compressed image anymore.
So I would suggests adding an option to define an archive file (similar to youtube-dl) which store the path of all images, which where already compressed. This would make it possible to continue compressing where you have stopped.

I can implement this, if this feature is wanted.

**Comment by andrews05** [comment]
Hm, I understand your issue but this seems like an incredibly niche feature which I would say falls out of scope of what oxipng should be providing.

Can you explain your scenario a little more? Like, could you just run oxipng on a subset of your files at a time so you know where you're up to?

**Comment by LuckyTurtleDev** [comment]
I have run oxipng at my home folder yesterday, but it has take forever and I have abort it. Today I  started it again, but it starts from the beginning. Because I have absolute no idea where it has stop I can not continue without starting from scratch. (It would also be a pain to define this per argument).

Some thing like `oxipng --archive foo.txt .` would allow to save a list of compressed files and to skip them if I execute the same command again.

> could you just run oxipng on a subset of your files at a time so you know where you're up to?

I think this should also be possible. But it much more complicated and annoying . You do not know how long this would take and if you abort the process you still don not know where you have stop. You must also note/remember which files you have compress before.

**Comment by ace-dent** [comment]
> could you just run oxipng on a subset of your files

I would say this is how *all* png optimizers work. Handling of bulk jobs is normally handled by a script / coded GUI, etc. ... Not sure it should be a core function here, due to portability and maintaining that extra code...

**Comment by andrews05** [comment]
Yeah, I hear the frustration but you should be able to script something yourself outside of oxipng. E.g. you could start by constructing a list of all your png files and then making a script to work through them. Or even simpler, you could try only finding files modified before a given date, so as to exclude ones that were recently optimised (assuming Windows has some sort of `find` equivalent that can do this?).

I'm afraid I just don't agree that this should be built-in to oxipng. That is just my opinion though - @AlexTMjugador or @shssoichiro may have different ideas.

**Comment by AlexTMjugador** [comment]
I agree with @andrews05 and @ace-dent that this feature is out of scope for OxiPNG. In my opinion, tracking the progress of a batch optimization job is better suited to a script or higher level application with more context about the job, and I find the alternative approaches suggested by @andrews05 to be spot on.

> I would suggests adding an option to define an archive file (similar to youtube-dl) which store the path of all images, which where already compressed.

This statement made me curious as to why `youtube-dl` has such an option, however. After some digging, I found out that it was introduced to handle new videos being added to time-varying video sources (i.e., playlists and the like). [[1]](https://github.com/ytdl-org/youtube-dl/commit/c1c9a79c49e8656f3244744e6f4e336e47a03206) [[2]](https://github.com/ytdl-org/youtube-dl/commit/5a3efcd27c1262cc7132f7e1a092524b580788ee)

I think this is a substantially different scenario than that of OxiPNG: video sources can change at indeterminate times, and there are no good ways to keep track of changes to video sources outside of `youtube-dl`: `youtube-dl` is usually the only reasonable way to query those video sources, which often have proprietary interfaces that change all the time. On the contrary, OxiPNG CLI users can easily know when the input dataset changes, because the interface is the filesystem.

All in all, I think this feature would trade complexity on one side for another, in a way that I'm not sure OxiPNG would do better than a custom script. It's also not that hard to process a few PNG files at a time and record their names in a temporary file to resume the job if necessary.

**Comment by aaronfranke** [comment]
For anyone who wants this feature, feel free to grab and compile my branch: https://github.com/aaronfranke/oxipng/tree/cache-opt-files and then run `oxipng` with the `--cache` flag, which will cache the hashes of optimized files in your user cache folder, so you can re-run `oxipng` on the same files and it will skip already-optimized files.

**Comment by andrews05** [comment]
Just to comment further here:
If we did want to proceed with a feature like this it would require extensive consideration about all the different ways and reasons people might want to use it, including problems that might be encountered and how they could be resolved. We can already see that people have different ideas on how it could/should be done, and what works well for one person may not necessarily work well for someone else.

@aaronfranke It might be good if you could detail your specific use case here, just for reference.

I'll stress again though, that this is _not_ a direction we want to take oxipng in at this time.

**Comment by aaronfranke** [comment]
Let's say I have a folder of 100 PNG files. I may run `oxipng` on this folder, which takes a long time. Then, I add 10 more PNG files to this folder so that there are 110 files. If I want to re-optimize the folder, running `oxipng` on the whole folder again will re-optimize all 110 files. I would like to optimize the folder without repeating the work already done to optimize the first 100 images, so that it only needs to optimize the 10 new images.

**Comment by LuckyTurtleDev** [comment]
I personal would prefer if the cache file stored at a user definded place. So it can be moved together with the pngs.
For example '--cache-file cache.txt'

**Comment by andrews05** [comment]
> [edit] Keeping track of a file queue is a feature you might expect from a GUI program such as [FileOptimizer](https://nikkhokkho.sourceforge.io/static.php?page=FileOptimizer). Have you tried using this instead? If you like oxipng, perhaps you could ask the developer to add it as a plugin. [edit2] I forgot, this is blocked on apng support in oxipng, which will be available in the next release.

FileOptimizer now supports oxipng 🙂

**Comment by Summertime** [comment]
I think it would be practical for oxipng to track this as opposed to a wrapper script:

Oxipng can hash specifically the compression-relevant chunks, without consideration for additional metadata. If another tool changed the `tEXt` data, it shouldn't necessitate running the entire gauntlet of compression again.

Additionally, for a situation where `oxipng -omax a.png` followed by a later `oxipng --strip=safe -omax a.png`, a wrapper script would have issues, while oxipng can strip the metadata, check the hash of the compression-relevant chunks, and skip the compression step knowing that it has done its best priorly.

Also having some awareness of changes in results of various settings would make it easier for oxipng instead of a wrapper, e.g. `-o5` followed by a later `-o6` means only attempting `--zc 12 -f 3,4`, which is doable by a wrapper, however any changes (however unlikely) caused by version changes in oxipng or other libraries it pulls in (zopfli for example) would be unknown to the wrapper script.

I think implementing it using a KV database would be most practical, as using flat files (`$hashdir/$hash` per hash) would mean having a 4KB overhead for just storing a version number and trials attempted

**Comment by Taffer** [comment]
I was thinking about this in the context of running `oxipng` as a `pre-commit` hook; I've set this up for my game repos and website repos, for example. Generally this will only run `oxipng` over files that have been touched and are part of a new commit.

In theory, you could have a big commit where everything has been touched without actually modifying the image data… maybe someone is updating copyright notices or URLs in text chunks or something.

I've also got a mild obsession with desktop wallpaper, and I also want to keep everything as small as possible; after I've been downloading new PNG wallpapers, I run `oxipng` over them, usually via `oxipng --opt max --alpha --interlace off $(fd -e png)`. If I kept track of the last time I ran `oxipng`, I could use `fd`'s `--changed-within` option to select only the new files, assuming their modification times were set to when I downloaded them.

`oxipng` could add a couple of extra, small chunks so it would know to skip attempts at re-compressing everything:

- `oxII` - xxHash (XXH32 or XXH64) of the uncompressed image data
- `oxIO` - xxHash (XXH32) of the `oxipng` settings used to produce the output

When it starts to work on a file, `oxipng` could:

```pseudocode
if xxhash(settings) == `oxiIO` contents:
    if xxhash(uncompressed image data) == `oxII` contents:
        skip image
```

This would be a *Quality of Life* and *Good First Issue* type of change. I don't know Rust, but I'll do some mucking around with [xxhash-rust](https://crates.io/crates/xxhash-rust) to see if I can get anything useful going.

**Comment by ace-dent** [comment]
@Taffer 

> ... could add a couple of extra, small chunks ...

I work with small files ~few hundred bytes. Adding extra chunks is too much overhead for default behaviour. 
For *large* files it might make sense as an option... but stored using text chunks, as key & value.
AFAIK `PNGOUT` has a degree of self-detection, by setting certain flag  bits in the ZLIB header, which was fairly unique (according to the author). I think it was:

```text
FLEVEL (Compression level)
         These flags are available for use by specific compression
         methods.  The "deflate" method (CM = 8) sets these flags as
         follows:

            0 - compressor used fastest algorithm
            1 - compressor used fast algorithm
            2 - compressor used default algorithm
            3 - compressor used maximum compression, slowest algorithm

         The information in FLEVEL is not needed for decompression; it
         is there to indicate if recompression might be worthwhile.
```

So by setting to the least common value... it may be used to indicate Oxipng was the last processor... It is hit-n-miss but incurs no overhead. 🤷🏼

**Comment by Taffer** [comment]
This could be behind a command-line option, or attempt to be "smart" about it if it's on by default; maybe skip writing these extra chunks if the image data is under some threshold when compressed?

@ace-dent what size are your PNGs?

In my PR (assuming I figure out enough Rust to make it work!) I'll put it behind a new option (`--include-oxipng-fingerprint` or ???).

**Comment by ace-dent** [comment]
@Taffer - I am perhaps an extreme use case. You can see examples of my small png images [here](https://github.com/ace-dent/8x8.me/tree/main/docs/art) and [here](https://github.com/ace-dent/dex98/tree/main/art/png). 🔍

**Comment by TPS** [comment]
@Taffer There's a few conceptual thoughts that might be worth working through for your scenario:

1.  What platform's your desktop?  Many current 1s can use [modern image formats w/ better compression](https://github.com/oxipng/oxipng/issues/730#issuecomment-3409013299) (except in rare cases) than PNG possibly could.

2. As OxiPNG's devs have said above, [they don't want to include this in the main app b/c out-of-scope since it's better handled externally](https://github.com/oxipng/oxipng/issues/549#event-10152309570).  (So actually Idk why this RFE's still open? 🤷🏾‍♂️)

3.  @Aaronfranke's [mod](https://github.com/oxipng/oxipng/issues/549#issuecomment-1752178452) actually implements this in the most sensible manner possible for the spirit of the app.  Did you try that?  But I've no idea whether it's been kept up-to-date, which leads to.…

4. Most anytime there's (especially a major) update, somehow the devs find a way to improve compression.  Then what's the point?  To absolutely satisfy your mentioned obsession, you'd have to rerun OxiPNG on _everything_ after such, anyway.  (Actually, your proposal to add chunks defeats that purpose, too.)  Even better, you'd _really_ have rerun, e.g., [FileOptimizer](https://github.com/oxipng/oxipng/issues/326#issuecomment-711372708) (which actually has this built-in) afterwards everytime _anything_ in _its_ PNG toolchain updates to be absolutely _(currently)_ sure you're covered.

So what exactly's your goal?

**Comment by Taffer** [comment]
@ace-dent Those are awesomely tiny, edge cases like that are great for testing.

@TPS I prefer PNG because I know it's not patent-encumbered, and WebP is a kitchen-sink format where it's (IMHO) too easy for people to accidentally switch to lossy compression. @aaronfranke's mod would certainly work for what I want, which is the ability to run this *efficiently* via automation that happens more frequently than compression optimisations are discovered.

I guess my goal would be for @AlexTMjugador to have another option and pick one (or similar) to add to `oxipng`. 🤷

**Comment by andrews05** [comment]
@Taffer It's an interesting idea, conceptually similar to what @montehurd suggested in #628. I've actually had thoughts along these lines before (though I never expressed them):
An option something like `--optimized-marker='some value'` which would store the value given in an `oxIP` chunk. If the file already has an `oxIP` chunk that matches the same value then optimisation would be skipped.
This gives flexibility in exactly how the user wants to track it. They can choose to store some form of the optimisation parameters, the oxipng version, or even a blank value (which would have only 12 bytes overhead).

I could potentially be open to such a feature if it satisfies enough users. The obvious downside vs a local cache is the increase to file size.

**Comment by aaronfranke** [comment]
For anyone still interested in my branch, I've updated it today. https://github.com/aaronfranke/oxipng/tree/cache-opt-files
